tsl/src/remote/txn_resolve.c

```diff
@@ -108,9 +142,22 @@ remote_txn_heal_data_node(PG_FUNCTION_ARGS)
 
 	/*
 	 * Perform cleanup of all records if there are no unknown txns.
+	 *
+	 * In a heavily loaded system there's a possibility of ongoing transactions always being
+	 * present in which case we will never get a chance to clean up entries in "remote_txn"
+	 * table. So, we track healed gids in a list and delete those specific rows to keep the
+	 * "remote_txn" table from growing up indefinitely.
 	 */
-	if (list_length(unknown_txn_gid) == 0)
-		remote_txn_persistent_record_delete_for_data_node(foreign_server_oid);
+	if (list_length(unknown_txn_gids) == 0)
+		remote_txn_persistent_record_delete_for_data_node(foreign_server_oid, NULL);
```

Comment:
I think this wildcard removal might be broken because the internal Scanner uses a "SnapshotSelf" by default, which means the delete scan here can see new entries at this instant that weren't present in the scan of gids above. We should probably modify the scan functions to use a common snapshot set in the beginning of the function (or use the TransactionSnapshot) so that they have the same view of the metadata.

Reply:
Ok, good catch. Using `GetTransactionSnapshot()` now.